eventlet/hubs/__init__.py

```diff
@@ -11,6 +14,15 @@
 _threadlocal = threading.local()
 
 
+# order is important, get_default_hub returns first available from here
+builtin_hub_names = ('epolls', 'kqueue', 'poll', 'selects')
+builtin_hub_modules = tuple(importlib.import_module('eventlet.hubs.' + name) for name in builtin_hub_names)
```

Comment:
? I'm unaware of that convention. Is there a reference you could cite? I'm interested to learn.
The heuristic I tend to use is that when you know a sequence won't need to be mutated, (I have always assumed) a tuple is slightly cheaper. My exception to that is that a list of one element is slightly syntactically neater (and harder to mistake) than a tuple of one element.

Reply:
Most of the code I've seen uses lists to store arbitrary number of elements of the same type (or union of types) and tuples for fixed-size heterogeneous (semantically) collections so I can't point towards a single reference. The best I can come up with is the documentation for `typing.List` (can contain arbitrary number of elements of one type) and `typing.Tuple` (can also contain arbitrary number of elements of one type but the main use case is elements of specified types, fixed size:

https://docs.python.org/3/library/typing.html#typing.List
https://docs.python.org/3/library/typing.html#typing.Tuple

Reply:
@jstasiak First and most important, whatever final decision you (both) make here, thanks for arguments. It's clear you want to do better.

- pure technical argumentation, how it works without human bias: Tuple is immutable array. List is mutable array. Immutability is safety, therefore default choice unless otherwise required by task.

Since this is public "chat", *irrelevant note to recruiters: will work for food where all decisions are based only on logic, ignoring human factor*.

- Linked docs on `typing.List` does not speak of number of types. Perhaps they edited docs? I see this text on `typing.List`:
>    Generic version of list. Useful for annotating return types. To annotate arguments it is preferred to use abstract collection types such as Mapping, Sequence, or AbstractSet.
- I like this sentence from docs on `typing.Tuple`, it is IMHO, very unbiased of number of types.
> A plain Tuple is equivalent to Tuple[Any, ...], and in turn to tuple.
- `typing.List/Tuple` is related but does not define `list/tuple` semantics, if anything, I think it should work backwards, because `typing` is used to assist by documenting/type checking code; as opposed to typed languages (irrelevant point - Python is untyped language) where declared types actually define semantics of code and type documentation/checking is a byproduct

Reply:
I believe the types documented by https://github.com/python/typeshed illustrate how tuples are used in the ecosystem (I'm testing https://github.com/python/typeshed/commit/5cda4224fe39564249c2e996d874b4e853898a31; the stats aren't precise but they're good enough, it's mostly standard library but also some third-party packages):

```
# fixed-size tuples like: (str, bytes) or (int, int, int)
typeshed% ag 'Tuple\[[^]]*, ?[^.]*\]' | wc -l
    1108

# verying-length single-type tuples used like immutable lists
typeshed% ag 'Tuple\[[^]]*, ?\.\.\.\]' | wc -l
     258
```

That said this is such a minor thing that I actually regret bringing it up.